File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        # Logic: No Clue -> Just perform BFS
        gene_banks_map = set(bank)

        def graph_bfs(curr_gene):
            genes_queue = []

            genes_queue.append((curr_gene, 0))
            visited_set = set()

            min_mutations = -1
            while len(genes_queue):
                curr_gene, curr_mutation = genes_queue.pop(0)

                if curr_gene == endGene:
                    min_mutations = curr_mutation
                    break

                if curr_gene not in visited_set:
                    visited_set.add(curr_gene)

                    for char_idx in range(len(curr_gene)):
                        for nucleotide in 'ACGT':
                            if curr_gene[char_idx] != nucleotide:
                                next_gene = curr_gene[:char_idx] + nucleotide + curr_gene[char_idx + 1:]
                                if next_gene in gene_banks_map and next_gene not in visited_set:
                                    genes_queue.append((next_gene, curr_mutation + 1))
            
            return min_mutations

        min_mutations = graph_bfs(startGene)

        return min_mutations

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.java
class Pair<G,M>{
    G currGene;
    M currMutations;

    Pair(G currGene, M currMutations){
        this.currGene = currGene;
        this.currMutations = currMutations;
    }
}

class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        HashSet<String> genesBankSet = new HashSet<>();

        for(String currGene : bank){
            genesBankSet.add(currGene);
        }

        HashSet<String> visitedSet = new HashSet<>();
        Deque<Pair<String, Integer>> genesQueue = new ArrayDeque<>();

        genesQueue.addLast(new Pair<>(startGene, 0));
        int totalMutations = -1;
        visitedSet.add(startGene);

        while(genesQueue.size() > 0){
            Pair<String, Integer> firstEl = genesQueue.removeFirst();

            String currGene = firstEl.currGene;
            int currMutations = firstEl.currMutations;

            if(currGene.equals(endGene)){
                totalMutations = currMutations;
                break;
            }

            for(int idxI = 0; idxI < currGene.length(); idxI++){
                for(char currChar : new char[] {'A','C','G','T'}){

                    String newGene = currGene.substring(0, idxI) + currChar + currGene.substring(idxI + 1);
                    if(!visitedSet.contains(newGene) && genesBankSet.contains(newGene)){
                        visitedSet.add(newGene);
                        genesQueue.addLast(new Pair<>(newGene, currMutations + 1));
                    }
                }
            }
        }

        return totalMutations;
    }
}